druid-shell/src/scale.rs

```diff
@@ -0,0 +1,388 @@
+        Scale {
+            dpi_x: SCALE_TARGET_DPI,
+            dpi_y: SCALE_TARGET_DPI,
+            scale_x: 1.0,
+            scale_y: 1.0,
+        }
+    }
+}
+
+impl Scale {
+    /// Create a new `Scale` state based on the specified DPIs.
+    ///
+    /// Use this constructor if the platform provided DPI is the most accurate number.
+    pub fn from_dpi(dpi_x: f64, dpi_y: f64) -> Scale {
+        Scale {
+            dpi_x,
+            dpi_y,
+            scale_x: dpi_x / SCALE_TARGET_DPI,
+            scale_y: dpi_y / SCALE_TARGET_DPI,
+        }
+    }
+
+    /// Create a new `Scale` state based on the specified scale factors.
+    ///
+    /// Use this constructor if the platform provided scale factor is the most accurate number.
+    pub fn from_scale(scale_x: f64, scale_y: f64) -> Scale {
+        Scale {
+            dpi_x: SCALE_TARGET_DPI * scale_x,
+            dpi_y: SCALE_TARGET_DPI * scale_y,
+            scale_x,
+            scale_y,
+        }
+    }
+
+    /// Returns the x axis platform DPI associated with this `Scale`.
+    #[inline]
+    pub fn dpi_x(&self) -> f64 {
+        self.dpi_x
+    }
+
+    /// Returns the y axis platform DPI associated with this `Scale`.
+    #[inline]
+    pub fn dpi_y(&self) -> f64 {
+        self.dpi_y
+    }
+
+    /// Returns `true` if the specified DPI is approximately equal to the `Scale` DPI.
+    #[inline]
+    pub fn dpi_approx_eq(&self, dpi_x: f64, dpi_y: f64) -> bool {
+        self.dpi_x.approx_eq(dpi_x, (f64::EPSILON, 2))
+            && self.dpi_y.approx_eq(dpi_y, (f64::EPSILON, 2))
+    }
+
+    /// Returns the x axis scale factor.
+    #[inline]
+    pub fn scale_x(&self) -> f64 {
+        self.scale_x
+    }
+
+    /// Returns the y axis scale factor.
+    #[inline]
+    pub fn scale_y(&self) -> f64 {
+        self.scale_y
+    }
+
+    /// Converts from display points into pixels, using the x axis scale factor.
+    #[inline]
+    pub fn dp_to_px_x<T: Into<f64>>(&self, x: T) -> f64 {
+        x.into() * self.scale_x
+    }
+
+    /// Converts from display points into pixels, using the y axis scale factor.
+    #[inline]
+    pub fn dp_to_px_y<T: Into<f64>>(&self, y: T) -> f64 {
+        y.into() * self.scale_y
+    }
+
+    /// Converts from display points into pixels,
+    /// using the x axis scale factor for `x` and the y axis scale factor for `y`.
+    #[inline]
+    pub fn dp_to_px_xy<T: Into<f64>>(&self, x: T, y: T) -> (f64, f64) {
+        (x.into() * self.scale_x, y.into() * self.scale_y)
+    }
+
+    /// Converts the `item` from display points into pixels,
+    /// using the x axis scale factor for coordinates on the x axis
+    /// and the y axis scale factor for coordinates on the y axis.
+    #[inline]
+    pub fn to_px<T: Scalable>(&self, item: &T) -> T {
+        item.to_px(self)
+    }
+
+    /// Converts from pixels into display points, using the x axis scale factor.
+    #[inline]
+    pub fn px_to_dp_x<T: Into<f64>>(&self, x: T) -> f64 {
```

Comment:
do we anticipate these two methods be used? If it's an obscure case, I would be happy to require the user to construct a point, convert that, and then take the `x`.